File: P54 Spiral Matrix /Solution.java
class Solution {
    // Time Complexity: O(N), where N is the total number of elements in the input matrix. We add every element in the matrix to our final answer.
    // Space Complexity: O(N), the information stored in res.
    public List<Integer> spiralOrder(int[][] matrix) {
        // 0,0 -> 0,1 -> 0,matrix[0].length -1 -> 
        // 1, matrix[0].length -1 -> .. -> matrix[1].length -1, matrix[0].length -1 ->
        // matrix[1].length -1, 1 -> matrix[1].length -1, matrix[0].length -1 -1 -> 
        
        List<Integer> res = new ArrayList<Integer>();
        
        if (matrix.length == 0) {
            return res;
        }
        
        int rowBegin = 0;
        int rowEnd = matrix.length-1;
        int colBegin = 0;
        int colEnd = matrix[0].length - 1;
        
        while (rowBegin <= rowEnd && colBegin <= colEnd) {
            // Traverse Right
            for (int j = colBegin; j <= colEnd; j ++) {
                res.add(matrix[rowBegin][j]);
            }
            rowBegin++;
            
            // Traverse Down
            for (int j = rowBegin; j <= rowEnd; j ++) {
                res.add(matrix[j][colEnd]);
            }
            colEnd--;
            
            if (rowBegin <= rowEnd) {
                // Traverse Left
                for (int j = colEnd; j >= colBegin; j --) {
                    res.add(matrix[rowEnd][j]);
                }
            }
            rowEnd--;
            
            if (colBegin <= colEnd) {
                // Traver Up
                for (int j = rowEnd; j >= rowBegin; j --) {
                    res.add(matrix[j][colBegin]);
                }
            }
            colBegin ++;
        }
        
        return res;
    }
}

File: P54 Spiral Matrix /solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []
        if not matrix:
            return res
        
        row_begin = 0
        row_end = len(matrix)-1
        col_begin = 0
        col_end = len(matrix[0]) - 1
        
        while row_begin <= row_end and col_begin <= col_end:
            # Traverse Right
            for i in range(col_begin, col_end+1):
                res.append(matrix[row_begin][i])
            row_begin += 1
            
            # Traverse Down
            for i in range(row_begin, row_end+1):
                res.append(matrix[i][col_end])
            col_end -= 1
            
            if row_begin <= row_end:
                # Traverse Left
                for i in range(col_end, col_begin-1, -1):
                    res.append(matrix[row_end][i])
            row_end -= 1
            
            if col_begin <= col_end:
                # Traver Up
                for i in range(row_end, row_begin-1, -1):
                    res.append(matrix[i][col_begin])
            col_begin += 1
            
        return res